Starting version of speech to text robots

peetucket opened the pull request:
## Why was this change made? 🤔

Part of #1341 

1. Renames "caption" robots to "speechToText" robots (workflow definition itself in workflow-server-rails still to come)
2. Adds likely candidates for other new robots will we need (analogous to ocrWF)
3. Create new class to handle opening an object with retries since we will want to do this for both ocr and speech to text
4. Use this new class in existing ocr robots and new speechToText robots to open objects.
5. Create a new class which will be used to determine if speech to text is needed (analogous to existing class for OCR). 
6. New tests

Workflow definition in https://github.com/sul-dlss/workflow-server-rails/pull/804

## How was this change tested? 🤨

Unit tests

peetucket commented on lib/dor/text_extraction/ocr.rb:
```diff
@@ -127,7 +127,7 @@ def cleanup_abbyy_exceptions
       end
 
-      # iterate through cocina strutural contains and return all File objects for files that need to be OCRed
+      # iterate through cocina structural contains and return all File objects for files that need to be OCRed
```
typo

peetucket commented on lib/dor/text_extraction/ocr.rb:
```diff
@@ -160,8 +160,7 @@ def allowed_mimetypes
 
-      # maps the allowed content types to the resource type we will look for files in
-      # TODO: refine list of allowed object types for OCR
+      # maps the allowed object types to the resource type we will look for files in
```
remove the 'todo' for OCR ... we are done for now

peetucket commented on lib/dor/text_extraction/version_updater.rb:
```diff
@@ -0,0 +1,41 @@
+      # rubocop:enable Metrics/MethodLength
+    end
+  end
+end
```
new class to handle object opening with retries (extracted from start-ocr robot so we can use in start-stt robot)

peetucket commented on lib/dor/text_extraction/speech_to_text.rb:
```diff
@@ -0,0 +1,95 @@
+      end
+    end
+  end
+end
```
analogous to ocr class to handle various speech to text related tasks (put in some sensible defaults, we can edit later on implementation)

peetucket commented on spec/robots/dor_repo/ocr/start_ocr_spec.rb:
```diff
@@ -36,41 +36,6 @@
-      end
-    end
-  end
-
```
these tests are no longer needed here and are instead performed in the new spec for the new `VersionUpdated` class

edsu reviewed (approved):
This all looks good. I left a comment that the approach to mime type matching may need to be generalized. I found myself kind of wanting to expand `stt` to `speech_to_text` and `SpeechToText` in places, but I can live with it being abbreviated.

peetucket commented on lib/dor/text_extraction/speech_to_text.rb:
```diff
@@ -0,0 +1,95 @@
+
+      # defines the mimetypes types for which speech to text files can possibly be run
+      # preferentially select files for speech to text by ordering of mimetypes below
+      # TODO: refine list of allowed mimetypes for speech to text
```
Yeah, good question.  I'd suggest we adjust later when we know better that types of formats the sul-embed media player will want.  I took this starter set from actual media objets, so they are at least somewhat real.

jmartin-sul replied:
i'd be fine with adjusting later.  maybe worth a backlog ticket for coming back to re-examine later?  or will this be obvious enough for us to not forget?

peetucket replied:
Creating a ticket so we don't forget.

peetucket commented:
> This all looks good. I left a comment that the approach to mime type matching may need to be generalized. I found myself kind of wanting to expand `stt` to `speech_to_text` and `SpeechToText` in places, but I can live with it being abbreviated.
> 
> I was wondering if it might be worth doing an integration test before merging?

Deploying this and running the OCR integration tests is a good idea.  I will do this before merging.

I am not ambivalent about the "stt" stuff - that's in the individual steps within the workflow itself, so it's was mostly to save space (it should be more clear what we are talking about given the workflow name is fully descriptive)

jmartin-sul commented:
> > This all looks good. I left a comment that the approach to mime type matching may need to be generalized. I found myself kind of wanting to expand `stt` to `speech_to_text` and `SpeechToText` in places, but I can live with it being abbreviated.
> > I was wondering if it might be worth doing an integration test before merging?
> 
> Deploying this and running the OCR integration tests is a good idea. I will do this before merging.
> 
> I am not ambivalent about the "stt" stuff - that's in the individual steps within the workflow itself, so it's was mostly to save space (it should be more clear what we are talking about given the workflow name is fully descriptive)

i also found myself sort of wanting `stt` expanded to prevent future confusion, but it's a good point that all the uses occur inside of `SpeechToText` (for the robot code) or `speechToTextWF` (for the workflow xml).  so i think that makes me neutral on that point.  either way, if the change were made, i'd lean slightly towards doing that in a follow-on PR, since it'd be easier to review separately.  or at least as a separate commit (given that it'd affect most or all of the files in this PR, which is necessarily large, and a bit unweildy to review).

jmartin-sul reviewed (approved):
i had a small number of nitpicky questions/suggestions, but i think i'd rather just merge this and do those in a follow up PR, if you think they're worthwhile.

agreed that an integration test before merging is a good idea, given the refactor of existing functionality (though the refactor LGTM based just on reading through).

peetucket commented on spec/robots/dor_repo/speech_to_text/start_stt_spec.rb:
```diff
@@ -0,0 +1,75 @@
+    let(:workflow_client) { instance_double(Dor::Workflow::Client) }
+    # NOTE: this is just mocking a workflow response, it doesn't actually have to be fully accurate
+    # we just need to expect we will update the status of all steps except start-stt
+    let(:skip_all) { instance_double(Dor::Workflow::Response, name: 'skip-all') }
```
it's actually used in a `before` block just below

peetucket commented:
> i also found myself sort of wanting `stt` expanded to prevent future confusion, but it's a good point that all the uses occur inside of `SpeechToText` (for the robot code) or `speechToTextWF` (for the workflow xml). so i think that makes me neutral on that point. either way, if the change were made, i'd lean slightly towards doing that in a follow-on PR, since it'd be easier to review separately. or at least as a separate commit (given that it'd affect most or all of the files in this PR, which is necessarily large, and a bit unweildy to review).

Integration test done and works fine for OCR.

Will leave any potential renaming from `stt` to `speech_to_text` to a future PR.

peetucket commented on spec/lib/dor/text_extraction/speech_to_text_spec.rb:
```diff
@@ -0,0 +1,108 @@
+
+  describe '#possible?' do
+    context 'when the object is not a DRO' do
+      let(:cocina_object) { instance_double(Cocina::Models::DRO, externalIdentifier: druid, dro?: false, type: object_type) }
```
First off, thank for the Simpsons reference (actually had to look that one up and I'm a Simpsons fan!).  And yes, making the double a `Collection` is less confusing and works just fine.  I'll change here and in the similar OCR test.

peetucket commented on lib/dor/text_extraction/version_updater.rb:
```diff
@@ -0,0 +1,41 @@
+        @max_tries = max_tries
+      end
+
+      def self.open(druid:, object_client:, description: '', max_tries: 3)
```
Fair enough.  We only have a couple callers and they provide descriptions, so no need for a default.

peetucket commented:
I believe I addressed all of the comments (except for the potential renaming, which is left to future work).  Will wait a bit to see if anyone wants to re-check, and will then merge if no objections.
